src/connector/src/sink/file_sink/batching_log_sink.rs

```diff
@@ -0,0 +1,145 @@
+                match &state {
+                    LogConsumerState::BarrierReceived { .. } => {
+                        // we need to force to finish the batch here. Otherwise, there can be data loss because actor can be dropped and rebuilt during scaling.
+                        if let Some(committed_chunk_id) = sink_writer.should_finish().await? {
+                            // When the chunk we need to truncate encounters an epoch increase, we need to truncate the barrier first.
+                            log_reader.truncate(TruncateOffset::Barrier { epoch: (epoch) })?;
+
+                            log_reader.truncate(TruncateOffset::Chunk {
+                                epoch: (epoch),
+                                chunk_id: (committed_chunk_id),
+                            })?
+                        };
+                    }
+                    _ => unreachable!(
+                        "update vnode bitmap can be accepted only right after \
+                    barrier, but current state is {:?}",
+                        state
+                    ),
+                }
+            }
+            // begin_epoch when not previously began
+            state = match state {
+                LogConsumerState::Uninitialized => {
+                    LogConsumerState::EpochBegun { curr_epoch: epoch }
+                }
+                LogConsumerState::EpochBegun { curr_epoch } => {
+                    assert!(
+                        epoch >= curr_epoch,
+                        "new epoch {} should not be below the current epoch {}",
+                        epoch,
+                        curr_epoch
+                    );
+                    LogConsumerState::EpochBegun { curr_epoch: epoch }
+                }
+                LogConsumerState::BarrierReceived { prev_epoch } => {
+                    assert!(
+                        epoch > prev_epoch,
+                        "new epoch {} should be greater than prev epoch {}",
+                        epoch,
+                        prev_epoch
+                    );
+                    LogConsumerState::EpochBegun { curr_epoch: epoch }
+                }
+            };
+            match item {
+                LogStoreReadItem::StreamChunk { chunk, chunk_id } => {
+                    match sink_writer
+                        .write_batch_and_try_finish(chunk, chunk_id)
+                        .await
+                    {
+                        Err(e) => {
+                            return Err(e);
+                        }
+                        // The file has been successfully written and is now visible to downstream consumers.
+                        // Truncate the file to remove the specified `chunk_id` and any preceding content.
+                        Ok(Some(chunk_id)) => {
+                            // When the chunk we need to truncate encounters an epoch increase, we need to truncate the barrier first.
```

Comment:
Ditto. The implementation doesn't match with the documentation here. In the current codes, we truncate barrier all the time, which is inccorrect.
- We only need to truncate barrier when epoch increases.
- We don't need to truncate chunk and barrier together on the same epoch. Even if we do, we should truncate chunk first.

Reply:
I misunderstand yiming's comment before. Assume there are two arrival orders:
1. `chunk_1, chunk_2, barrier_1` and `chunk_1` is written 
We should **not** truncate barrier 1 but only truncate chunk_1, as truncate barrier1 will make chunk_2 removed in log store.
2. `chunk_1, barrier_1, chunk_2,` and `chunk_2` is written, this is the case you and yiming mentioned:
We should first truncate barrier_1 and then truncate chunk_2.

Based on above, when we need to truncate a chunk, we need 
- firstly check whether current_epoch > prev_epoch, and truncate prev_epoch if epoch increase
- secondly truncate chunk.